libcodechecker/log/log_parser.py

```diff
@@ -0,0 +1,791 @@
+    pointing to a flag to replace.
+    """
+    value = REPLACE_OPTIONS_MAP.get(flag_iterator.item)
+
+    if value:
+        buildaction.analyzer_options.extend(value)
+
+    return bool(value)
+
+
+def __skip(flag_iterator, buildaction):
+    """
+    This function skips the flag pointed by the given flag_iterator with its
+    parameters if any.
+    """
+    if IGNORED_OPTIONS.match(flag_iterator.item):
+        return True
+
+    for pattern, arg_num in IGNORED_PARAM_OPTIONS.items():
+        if re.match(pattern, flag_iterator.item):
+            for i in range(arg_num):
+                next(flag_iterator)
+            return True
+
+    return False
+
+
+def parse_options(compilation_db_entry):
+    """
+    This function parses a GCC compilation action and returns a BuildAction
+    object which can be the input of Clang analyzer tools.
+
+    compilation_db_entry -- An entry from a valid compilation database JSON
+                            file, i.e. a dictionary with the compilation
+                            command, the compiled file and the current working
+                            directory.
+    """
+
+    buildaction = BuildAction()
+
+    if 'arguments' in compilation_db_entry:
+        gcc_command = compilation_db_entry['arguments']
+        buildaction.original_command = ' '.join(gcc_command)
+    elif 'command' in compilation_db_entry:
+        buildaction.original_command = compilation_db_entry['command']
+        gcc_command = compilation_db_entry['command'] \
+            .replace(r'\"', '"') \
+            .replace(r'"', r'"\"')
+        gcc_command = shlex.split(gcc_command)
+    else:
+        raise KeyError("No valid 'command' or 'arguments' entry found!")
+
+    buildaction.directory = compilation_db_entry['directory']
+    buildaction.action_type = buildaction.COMPILE
+    buildaction.compiler = gcc_command[0]
+    if '++' in buildaction.compiler or 'cpp' in buildaction.compiler:
+        buildaction.lang = 'c++'
+
+    flag_transformers = [
+        __get_arch,
+        __get_language,
+        __get_output,
+        __determine_action_type,
+        __replace,
+        __skip,
+        __append_from_file,
+        __collect_compile_opts,
+        __collect_sources]
+
+    for it in OptionIterator(gcc_command[1:]):
+        for flag_transformer in flag_transformers:
+            if flag_transformer(it, buildaction):
+                break
+        else:
+            pass
+            # print('Unhandled argument: ' + it.item)
+
+    def abs_path_maker(p):
+        return os.path.normpath(
+            os.path.join(compilation_db_entry['directory'], p))
+
+    merge_space_files(buildaction.sources, compilation_db_entry['file'])
+    buildaction.sources = map(abs_path_maker, buildaction.sources)
+
+    compiled_file = abs_path_maker(compilation_db_entry['file'])
+    if compiled_file != '.' and \
+            compiled_file not in map(abs_path_maker, buildaction.sources):
```

Comment:
Why do we convert the paths to absolute twice?
Also, how is it possible to not have `compiled_file` in the `buildaction.source`? Shouldn't this be a sanity check rather than part of the logic?

Reply:
It was really a mistake to convert these paths to absolute twice.
The use-case of this check was that in case of source files containing space in their names, it was impossible to determine whether those are two separate arguments or a filename with space in it. We also have a test case on this. Nonetheless, I'll check if it is still relevant.